contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c

```diff
@@ -374,6 +374,228 @@ tsql_check_param_readonly(const char *paramname, TypeName *typename, bool readon
 	}
 }
 
+
+/*
+* This function takes a JsonExpression, and an optional path then
+* calls the openjson_simple function
+*/
+Node *
+TsqlOpenJSONSimpleMakeFuncCall(Node* jsonExpr, Node* path)
+{
+    FuncCall *fc;
+    if(path)
+    {
+        fc = makeFuncCall(TsqlSystemFuncName("openjson_simple"), list_make2(jsonExpr, path), COERCE_EXPLICIT_CALL, -1);
+    }
+    else
+    {
+        fc = makeFuncCall(TsqlSystemFuncName("openjson_simple"), list_make1(jsonExpr), COERCE_EXPLICIT_CALL, -1);
+    }
+    return (Node*) fc;
+}
+
+/*
+* This function takes a JsonExpression, path, column list, and optional alias
+* It acts as a bridge between the parser and the json_with function by correctly
+* assembling the function arguments, column definitions list, and alias
+*/
+Node *
+TsqlOpenJSONWithMakeFuncCall(Node* jsonExpr, Node* path, List* cols, Alias* alias)
+{
+    FuncCall *fc;
+    List *jsonWithParams = list_make2(jsonExpr, path);
+    ListCell *lc;
+    RangeFunction *rf = makeNode(RangeFunction);
+    Alias *a = makeNode(Alias);
+    a->aliasname = alias != NULL ? alias->aliasname : "f";
+
+    foreach(lc, cols)
+    {
+        OpenJson_Col_Def *cd = (OpenJson_Col_Def*) lfirst(lc);
+        int initialTmod = getElemTypMod(cd->elemType);
+        char* typeNameString = TypeNameToString(cd->elemType);
+        ColumnDef *n = (ColumnDef *) createOpenJsonWithColDef(cd->elemName, cd->elemType);
+        StringInfo format_cols = makeStringInfo();
+
+        if(strcmp(cd->elemPath, "") == 0)
+        {
+            // If not path is provided with use the standard path [$.columnName]
+            appendStringInfo(format_cols, "$.%s ", cd->elemName);
+        }
+        else
+        {
+            appendStringInfo(format_cols, "%s ", cd->elemPath);
+        }
+
+        // character types need to have the typmod appended to them
+        if(isCharType(typeNameString))
+        {
+            int newTypMod = getElemTypMod(n->typeName);
```

Comment:
nit; can we just use `initialTmod` here?

Reply:
We change the underlying typmod in the case that none is provided (set the default to 1) or the case that its max (the underlying representation is -8000 so we change to the proper representation of 4000). We get the initial typmod before the change so we are able to check if it was MAX for nvarchar(max) as json